Automattic/wp-calypso
# Plans: Launch unlaunched site when upgrading to atomic plan

Note: This is currently targeting the branch in #34193. If that lands first, retarget on master.

As shown in #32671, sites should be launched when upgrading to a plan which takes the site Atomic (at least until we have a good analog for private / unlaunched sites on Atomic).

This change leverages the checkout lib to do just that. It also adds a `Dialog` component to inform the customer that upgrading to plans which will take a site Atomic will make them publicly accessible:

![upgrade-to-atomic-set-to-public](https://user-images.githubusercontent.com/1587282/62673955-524f9880-b96e-11e9-9935-1d17511ea196.gif)

#### Changes proposed in this Pull Request

* Add a `Dialog` signifying that selecting certain plans will immediately launch the site
* Clicking the primary button of the Dialog redirects to the checkout screen
* Set the site to public when a plan which takes the site Atomic is purchased as an upgrade if the site is private
* Add a `QuerySites` component to the Plan features component (so we can reliably check if the site is private / unlaunched)
* Make `isUnlaunchedSite` selector return `false` when no site is selected (to match function documentation saying it should return a bool)


#### Testing instructions

##### Upgrade a Private by default site

* Run this branch
* Browse to `/start`
* Mouse over the "environment badge"
* Mouse over `ABTESTS`
* Scroll down the list of active tests until you see `privateByDefault`
* Click the radio button for `selected`
* Create a Free site (any segment except for online-store)
  * It should be unlaunched ("private by default")
* Go to the plans page
* Click `Business Sites and Online Stores`
* Click the `Upgrade` button for the `Business` or `Online Store` plan
  * You should see a Dialog informing you that your site will be public upon upgrade to this plan
    * Clicking `Cancel` should return you to the plans listing
    * Clicking `Let's Do It` should take you to checkout
      * The appropriate plan should be in your cart
      * You should be able to complete checkout
        * The appropriate upgrades should be in place
* Repeat the above and try to break the flow -- report cases where you can!

##### Upgrade a Private site that's not PbD

* Run this branch
* Browse to `/start`
* Mouse over the "environment badge"
* Mouse over `ABTESTS`
* Scroll down the list of active tests until you see `privateByDefault`
* Click the radio button for `control`
* Create a Free site (any segment except for online-store)
  * It should be public (**NOT** "private by default")
* Change the site to Private
  * `Manage | Settings | Privacy` => `Private`
* Repeat the above -- it should behave identically

##### Upgrade a Public site

* Repeat the previous section, but do not set the site to private -- make sure it's public before the upgrade
* Complete the upgrade to a `Business` or `Online Store` plan
  * You should not see the `Dialog` & the site should remain public

##### Purchase Plans During Signup

* Complete the `/start` signup flow & choose a `Business` or `Online Store` plan
  * You should not see the `Dialog` (for neither email confirmation, nor going public)
  * Your site should be Public after checkout
* Repeat the above -- all other plans should be Private by Default & no `Dialog`s should be shown during signup

##### Test out Jetpack plan purchase & upgrade flows

* Jetpack plan flows should *NOT* see the `Dialog`s & should not otherwise be affected by this change.

Fixes #32671
